### PR TITLE
Decouple Cordova from PhoneGap.

### PR DESCRIPTION
**Pull request by stevengill**
This PR will replace #650.

### CODE REVIEW

File: spec/phonegap/cordova.spec.js [hunk 1]
```diff
@@ -136,63 +178,63 @@ describe('phonegap.cordova(options, [callback])', function() {
...
         });
 
         describe('when not a valid cordova project', function() {
             beforeEach(function() {
                 cordova.util.isCordova.andReturn(false);
                 shell.exec.andCallThrough();
             });
-
-            it('should trigger the callback without an error', function(done) {
+            // ToDo: @carynbear this not testable with current dependency implementation b/c no cordova to call without project.
```
Comment (filmaj) [inline spec/phonegap/cordova.spec.js]: this comment is confusing. so, is this testable or not?

File: spec/phonegap/create.spec.js [hunk 2]
```diff
@@ -70,183 +257,25 @@ describe('phonegap.create(options, [callback])', function() {
...
-                callback(new Error('path already exists'));
-            });
-        });
 
-        it('should trigger callback with an error', function(done) {
-            phonegap.create(options, function(e) {
-                expect(e).toEqual(jasmine.any(Error));
-                done();
-            });
+    it('should trigger callback with an error because cordova-create is not mocked', function(done) {
```
Comment (filmaj) [inline spec/phonegap/create.spec.js]: what's the purpose of the test here? the test title confuses me - is this testing the tests? why is mocking mentioned?

**Comment by filmaj** [comment]
The tests are also really slow, took about 15 seconds on my machine (compared to sub-second now). It looks like a set of a few tests are taking the most time. Are we sure we are mocking out any slow I/O out of these test runs?

**Comment by filmaj** [comment]
Re: slow tests, I think the `cli.spec.js` is what is slow, since that actually shells out. Running `jasmine-node` with `--verbose` confirms as much:

```
$ phonegap [options] commands - 10259 ms
    should support no arguments - 2131 ms
    should support commands - 2079 ms
    should support options - 2043 ms
    should have exit code 0 - 2053 ms

    on an error - 1953 ms
        should have non-zero exit code - 1953 ms
```

I had removed `cli.spec.js` in favour of `shell.spec.js` in an earlier commit - I recommend removing `cli.spec.js` in this PR, too, as we already have `shell.spec.js`: https://github.com/phonegap/phonegap-cli/commit/382fb923033493446e57c50b3b5e682a9d2552a7

File: spec/phonegap/cordova.spec.js (hunk 1 above)
Comment (filmaj) [inline spec/phonegap/cordova.spec.js]: This is also a slow test, as we are calling through down to `shell.exec`. We should revisit what this is testing, and how to test it.

File: package.json [hunk 3]
```diff
@@ -31,28 +31,31 @@
   "dependencies": {
     "archiver": "^1.0.0",
     "colors": "0.6.0-1",
-    "connect-phonegap": "0.24.5",
-    "cordova": "6.4.0",
+    "connect-phonegap": "https://github.com/stevengill/connect-phonegap.git#177",
+    "cordova-create": "^1.0.1",
+    "cordova-common": "^2.0.0",
+    "elementtree": "^0.1.6",
```
Comment (filmaj) [inline package.json]: @stevengill didn't you say there was a breaking change in elementtree 0.1.7 that fucked shit up outta nowhere for you?

**Comment by filmaj** [comment]
OK, one more comment here: looks like we are adding a bunch of filesystem fixtures under `spec/testcases`. Do we need those? We don't want unit tests touching the filesystem at all. At a minimum, we don't need the cordova logo, css, etc. as part of the fixtures. Best case, we don't need them at all, and mock out all relevant filesystem interactions.

**Comment by surajpindoria** [comment]
I noticed this as well, took a quick glance and it seems that a lot of the fixtures are very similar. If they are needed maybe we could look into condensing them down.

File: spec/phonegap/cordova.spec.js (hunk 1 above)
Comment (stevengill) [inline spec/phonegap/cordova.spec.js]: yeah, comment is confusing

File: package.json (hunk 3 above)
Comment (stevengill) [inline package.json]: yup. For cordova-lib. Fixing that here now

File: spec/cli/create.spec.js [hunk 4]
```diff
@@ -278,29 +279,30 @@ describe('phonegap create <path>', function() {
...
-                id: undefined,
-                name: undefined,
-                config: {},
-                'copy-from': undefined,
-                'link-to': 'path/to/an-app'
-            },
-            jasmine.any(Function));
-        });
-    });
+    // describe('$ phonegap create ./my-app --link-to path/to/an-app', function() {
```
Comment (stevengill) [inline spec/cli/create.spec.js]: i believe we dropped support for --link-to when creating a project

File: spec/phonegap/create.spec.js (hunk 2 above)
Comment (stevengill) [inline spec/phonegap/create.spec.js]: Kind of useless. The old test forced the failure. The new one tries to run cordova-create under the hood with throws an error. If cordova-create was mocked like above, we could force the failure.

**Comment by filmaj** [comment]
I think if you pull this branch in: https://github.com/filmaj/phonegap-cli/tree/176-nofixtures

... we can remove the test fixtures under `spec/testcases`.

File: lib/phonegap/cordova-dependence.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                var deferred = Q.defer();
+                // Find the latest version of Cordova published and install it
+                var cordovaCommand = 'npm install cordova --save';
+                extEvents.emit('log', 'Adding Cordova dependency with: "' + cordovaCommand + '"');
+                extEvents.emit('log', 'This may take a few minutes ...');
+                var child = shell.exec(cordovaCommand, {
+                    silent: true
+                }, function(code, stdout, stderr) {
+                    if (code != 0) {
+                        e = new Error('Error from npm: ' + stdout.replace('\n', ''));
```
Comment (filmaj) [inline lib/phonegap/cordova-dependence.js]: Another way to frame this question: in the case that `npm install cordova --save` fails, is valuable output spit out in `stdout` or in `stderr` ?

File: spec/phonegap/create.spec.js (hunk 2 above)
Comment (filmaj) [inline spec/phonegap/create.spec.js]: I would remove this test for the purposes of this PR. We can always add future test coverage to, at a finer level, exercise the various failure modes.

File: spec/cli/create.spec.js (hunk 4 above)
Comment (filmaj) [inline spec/cli/create.spec.js]: Let's remove it then!

File: lib/phonegap/cordova-dependence.js (hunk 5 above)
Comment (stevengill) [inline lib/phonegap/cordova-dependence.js]: good point. I just did a test and the error propagates through stdout, not stderr...

**Pull request merged**